File: Dynamic_programming/deepDive/coinChange.ts
function coinChange(coins: number[], amount: number): number {
  let q: number[] = [];
  q.push(amount);
  let visited: Set<number> = new Set();
  visited.add(amount);
  let steps = 0;
  while (q.length > 0) {
    let size = q.length;
    for (let i = 0; i < size; i++) {
      let node = q.shift();
      if (node === undefined) continue;
      if (node === 0) return steps;
      for (let coin of coins) {
        let next = node - coin;
        if (!visited.has(next)) {
          visited.add(next);
          q.push(next);
        }
      }
    }
    steps++;
  }
  return -1;
}

console.log(coinChange([1, 2, 5], 11));

File: recursion/deepDive/coin_change.py
# https://leetcode.com/problems/coin-change/
"""
https://docs.google.com/document/d/1Uhnldz90UuZU5vDdVRAQgV49aKyR7ieafbmJTrVAH0U/edit?usp=sharing
11 = 1+1+....1 (11 tiems)
11 = 5+5+1 -> The minimum number of coins.
11 = 2+2+2+2+2 + 1
11 = 2+ 2+ 2 + 5
"""


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0

        if amount < 0:
            return -1

        result = float("inf")
        for coin in coins:
            branch = self.coinChange(coins, amount - coin)
            if branch != -1:
                result = min(result, branch + 1)

        return result if result != float("inf") else -1


# With memoization:

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @cache
        def f(amount: int) -> int:
            if amount == 0:
                return 0

            if amount < 0:
                return -1

            result = float("inf")
            for coin in coins:
                branch = f(amount - coin)
                if branch != -1:
                    result = min(result, branch + 1)
            return result

        result = f(amount)
        return result if result != float("inf") else -1
